Panic during compilation if function returns structure with unset slice

I run this code:

```go
package foo

type Foo struct {
	x []int
}

func NewFoo() Foo {
	return Foo{}
}

func Main() int {
	foo := NewFoo()
	if foo.x == nil {
		return 1
	}

	return 0
}
```

I expect to get `1` but got panic during compilation. If I use `return Foo{x:nil}` panic does not occur and I got correct result.

Panic message:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x8098fb]

goroutine 19 [running]:
testing.tRunner.func1.1(0x891a80, 0xcd5eb0)
	/snap/go/current/src/testing/testing.go:940 +0x2f5
testing.tRunner.func1(0xc0000d37a0)
	/snap/go/current/src/testing/testing.go:943 +0x3f9
panic(0x891a80, 0xcd5eb0)
	/snap/go/current/src/runtime/panic.go:969 +0x166
github.com/nspcc-dev/neo-go/pkg/compiler.(*codegen).emitLoadConst(0xc0000e9e00, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/alexvanin/Project/Neo/neo-go/pkg/compiler/codegen.go:115 +0x4b
github.com/nspcc-dev/neo-go/pkg/compiler.(*codegen).convertStruct(0xc0000e9e00, 0xc0000acc80)
	/home/alexvanin/Project/Neo/neo-go/pkg/compiler/codegen.go:1211 +0x25a
github.com/nspcc-dev/neo-go/pkg/compiler.(*codegen).Visit(0xc0000e9e00, 0x9c4be0, 0xc0000acc80, 0x0, 0x0)
	/home/alexvanin/Project/Neo/neo-go/pkg/compiler/codegen.go:568 +0x1382
go/ast.Walk(0x9c1780, 0xc0000e9e00, 0x9c4be0, 0xc0000acc80)
	/snap/go/current/src/go/ast/walk.go:52 +0x66
github.com/nspcc-dev/neo-go/pkg/compiler.(*codegen).Visit(0xc0000e9e00, 0x9c52a0, 0xc0000b7060, 0xc000099660, 0x0)
	/home/alexvanin/Project/Neo/neo-go/pkg/compiler/codegen.go:437 +0x33cf
go/ast.Walk(0x9c1780, 0xc0000e9e00, 0x9c52a0, 0xc0000b7060)
	/snap/go/current/src/go/ast/walk.go:52 +0x66
go/ast.walkStmtList(0x9c1780, 0xc0000e9e00, 0xc000099410, 0x1, 0x1)
	/snap/go/current/src/go/ast/walk.go:32 +0x9e
go/ast.Walk(0x9c1780, 0xc0000e9e00, 0x9c49e0, 0xc00009de00)
	/snap/go/current/src/go/ast/walk.go:224 +0x1adf
github.com/nspcc-dev/neo-go/pkg/compiler.(*codegen).convertFuncDecl(0xc0000e9e00, 0x9c4de0, 0xc0000ee180, 0xc00009de30)
	/home/alexvanin/Project/Neo/neo-go/pkg/compiler/codegen.go:263 +0x542
github.com/nspcc-dev/neo-go/pkg/compiler.(*codegen).compile(0xc0000e9e00, 0xc0000b73a0, 0xc0000c0370, 0xc0000a5230, 0x0)
	/home/alexvanin/Project/Neo/neo-go/pkg/compiler/codegen.go:1317 +0x490
github.com/nspcc-dev/neo-go/pkg/compiler.CodeGen(0xc0000b73a0, 0xc0000b6f20, 0xc0000b73a0, 0x0, 0x0, 0xc00009dc01, 0xc0000b6f20)
	/home/alexvanin/Project/Neo/neo-go/pkg/compiler/codegen.go:1345 +0x289
github.com/nspcc-dev/neo-go/pkg/compiler.CompileWithDebugInfo(0x9c1d60, 0xc0000b6f20, 0x7f8a774707d0, 0x0, 0x40e0e6, 0xc0000b6ee0, 0x20, 0x20)
	/home/alexvanin/Project/Neo/neo-go/pkg/compiler/compiler.go:78 +0x90
github.com/nspcc-dev/neo-go/pkg/compiler.Compile(...)
	/home/alexvanin/Project/Neo/neo-go/pkg/compiler/compiler.go:64
github.com/nspcc-dev/neo-go/pkg/compiler_test.vmAndCompileInterop(0xc0000d37a0, 0x92f6bc, 0xae, 0xc0000a5150, 0xc000044738)
	/home/alexvanin/Project/Neo/neo-go/pkg/compiler/vm_test.go:66 +0x10b
github.com/nspcc-dev/neo-go/pkg/compiler_test.vmAndCompile(...)
	/home/alexvanin/Project/Neo/neo-go/pkg/compiler/vm_test.go:56
github.com/nspcc-dev/neo-go/pkg/compiler_test.eval(0xc0000d37a0, 0x92f6bc, 0xae, 0x90bb00, 0xc0000b6ee0)
	/home/alexvanin/Project/Neo/neo-go/pkg/compiler/vm_test.go:33 +0x49
github.com/nspcc-dev/neo-go/pkg/compiler_test.TestForLoopRangeGlobalIndex1(0xc0000d37a0)
	/home/alexvanin/Project/Neo/neo-go/pkg/compiler/for_test.go:486 +0x7d
testing.tRunner(0xc0000d37a0, 0x936d68)
	/snap/go/current/src/testing/testing.go:991 +0xdc
created by testing.(*T).Run
	/snap/go/current/src/testing/testing.go:1042 +0x357

```

--- a/pkg/compiler/analysis.go
+++ b/pkg/compiler/analysis.go
@@ -44,8 +44,9 @@ func typeAndValueForField(fld *types.Var) (types.TypeAndValue, error) {
 		default:
 			return types.TypeAndValue{}, fmt.Errorf("could not initialize struct field %s to zero, type: %s", fld.Name(), t)
 		}
+	default:
+		return types.TypeAndValue{Type: t}, nil
 	}
-	return types.TypeAndValue{}, nil
 }
 
 // countGlobals counts the global variables in the program to add

--- a/pkg/compiler/codegen.go
+++ b/pkg/compiler/codegen.go
@@ -617,25 +617,31 @@ func (c *codegen) Visit(node ast.Node) ast.Visitor {
 			ast.Walk(c, n.X)
 			ast.Walk(c, n.Y)
 
-			switch {
-			case n.Op == token.ADD:
+			switch n.Op {
+			case token.ADD:
 				// VM has separate opcodes for number and string concatenation
 				if isStringType(tinfo.Type) {
 					emit.Opcode(c.prog.BinWriter, opcode.CAT)
 				} else {
 					emit.Opcode(c.prog.BinWriter, opcode.ADD)
 				}
-			case n.Op == token.EQL:
-				// VM has separate opcodes for number and string equality
-				op := c.getEqualityOpcode(n.X)
-				emit.Opcode(c.prog.BinWriter, op)
-			case n.Op == token.NEQ:
-				// VM has separate opcodes for number and string equality
-				if isStringType(c.typeInfo.Types[n.X].Type) {
-					emit.Opcode(c.prog.BinWriter, opcode.EQUAL)
-					emit.Opcode(c.prog.BinWriter, opcode.NOT)
+			case token.EQL, token.NEQ:
+				if isExprNil(n.X) || isExprNil(n.Y) {
+					c.prog.Err = errors.New("comparison with `nil` is not supported, use `len(..) == 0` instead")
+					return nil
+				}
+				if n.Op == token.EQL {
+					// VM has separate opcodes for number and string equality
+					op := c.getEqualityOpcode(n.X)
+					emit.Opcode(c.prog.BinWriter, op)
 				} else {
-					emit.Opcode(c.prog.BinWriter, opcode.NUMNOTEQUAL)
+					// VM has separate opcodes for number and string equality
+					if isStringType(c.typeInfo.Types[n.X].Type) {
+						emit.Opcode(c.prog.BinWriter, opcode.EQUAL)
+						emit.Opcode(c.prog.BinWriter, opcode.NOT)
+					} else {
+						emit.Opcode(c.prog.BinWriter, opcode.NUMNOTEQUAL)
+					}
 				}
 			default:
 				c.convertToken(n.Op)

--- a/pkg/compiler/codegen.go
+++ b/pkg/compiler/codegen.go
@@ -533,6 +533,8 @@ func (c *codegen) Visit(node ast.Node) ast.Visitor {
 			c.emitLoadConst(value)
 		} else if tv := c.typeInfo.Types[n]; tv.Value != nil {
 			c.emitLoadConst(tv)
+		} else if n.Name == "nil" {
+			c.emitDefault(new(types.Slice))
 		} else {
 			c.emitLoadLocal(n.Name)
 		}
@@ -1230,8 +1232,29 @@ func (c *codegen) convertStruct(lit *ast.CompositeLit) {
 
 		emit.Opcode(c.prog.BinWriter, opcode.DUP)
 		emit.Int(c.prog.BinWriter, int64(i))
-		c.emitLoadConst(typeAndVal)
+		c.emitDefault(typeAndVal.Type)
 		emit.Opcode(c.prog.BinWriter, opcode.SETITEM)
+	}
+}
+
+func (c *codegen) emitDefault(typ types.Type) {
+	switch t := c.scTypeFromGo(typ); t {
+	case "Integer":
+		emit.Int(c.prog.BinWriter, 0)
+	case "Boolean":
+		emit.Bool(c.prog.BinWriter, false)
+	case "String":
+		emit.String(c.prog.BinWriter, "")
+	case "Map":
+		emit.Opcode(c.prog.BinWriter, opcode.NEWMAP)
+	case "Struct":
+		emit.Int(c.prog.BinWriter, int64(typ.(*types.Struct).NumFields()))
+		emit.Opcode(c.prog.BinWriter, opcode.NEWSTRUCT)
+	case "Array":
+		emit.Int(c.prog.BinWriter, 0)
+		emit.Opcode(c.prog.BinWriter, opcode.NEWARRAY)
+	case "ByteArray":
+		emit.Bytes(c.prog.BinWriter, []byte{})
 	}
 }
 

--- a/pkg/compiler/codegen.go
+++ b/pkg/compiler/codegen.go
@@ -470,7 +470,6 @@ func (c *codegen) Visit(node ast.Node) ast.Visitor {
 	case *ast.SwitchStmt:
 		ast.Walk(c, n.Tag)
 
-		eqOpcode := c.getEqualityOpcode(n.Tag)
 		switchEnd, label := c.generateLabel(labelEnd)
 
 		lastSwitch := c.currentSwitch
@@ -490,7 +489,7 @@ func (c *codegen) Visit(node ast.Node) ast.Visitor {
 				for j := range cc.List {
 					emit.Opcode(c.prog.BinWriter, opcode.DUP)
 					ast.Walk(c, cc.List[j])
-					emit.Opcode(c.prog.BinWriter, eqOpcode)
+					c.emitEquality(n.Tag, token.EQL)
 					if j == l-1 {
 						emit.Jmp(c.prog.BinWriter, opcode.JMPIFNOT, lEnd)
 					} else {
@@ -630,19 +629,7 @@ func (c *codegen) Visit(node ast.Node) ast.Visitor {
 					c.prog.Err = errors.New("comparison with `nil` is not supported, use `len(..) == 0` instead")
 					return nil
 				}
-				if n.Op == token.EQL {
-					// VM has separate opcodes for number and string equality
-					op := c.getEqualityOpcode(n.X)
-					emit.Opcode(c.prog.BinWriter, op)
-				} else {
-					// VM has separate opcodes for number and string equality
-					if isStringType(c.typeInfo.Types[n.X].Type) {
-						emit.Opcode(c.prog.BinWriter, opcode.EQUAL)
-						emit.Opcode(c.prog.BinWriter, opcode.NOT)
-					} else {
-						emit.Opcode(c.prog.BinWriter, opcode.NUMNOTEQUAL)
-					}
-				}
+				c.emitEquality(n.X, n.Op)
 			default:
 				c.convertToken(n.Op)
 			}
@@ -988,13 +975,26 @@ func (c *codegen) getLabelOffset(typ labelOffsetType, name string) uint16 {
 	return c.labels[labelWithType{name: name, typ: typ}]
 }
 
-func (c *codegen) getEqualityOpcode(expr ast.Expr) opcode.Opcode {
+func (c *codegen) emitEquality(expr ast.Expr, op token.Token) {
 	t, ok := c.typeInfo.Types[expr].Type.Underlying().(*types.Basic)
-	if ok && t.Info()&types.IsNumeric != 0 {
-		return opcode.NUMEQUAL
+	isNum := ok && t.Info()&types.IsNumeric != 0
+	switch op {
+	case token.EQL:
+		if isNum {
+			emit.Opcode(c.prog.BinWriter, opcode.NUMEQUAL)
+		} else {
+			emit.Opcode(c.prog.BinWriter, opcode.EQUAL)
+		}
+	case token.NEQ:
+		if isNum {
+			emit.Opcode(c.prog.BinWriter, opcode.NUMNOTEQUAL)
+		} else {
+			emit.Opcode(c.prog.BinWriter, opcode.EQUAL)
+			emit.Opcode(c.prog.BinWriter, opcode.NOT)
+		}
+	default:
+		panic("invalid token in emitEqual()")
 	}
-
-	return opcode.EQUAL
 }
 
 // getByteArray returns byte array value from constant expr.

--- a/pkg/compiler/debug.go
+++ b/pkg/compiler/debug.go
@@ -183,7 +183,11 @@ func (c *codegen) scReturnTypeFromScope(scope *funcScope) string {
 }
 
 func (c *codegen) scTypeFromExpr(typ ast.Expr) string {
-	switch t := c.typeInfo.Types[typ].Type.(type) {
+	return c.scTypeFromGo(c.typeInfo.Types[typ].Type)
+}
+
+func (c *codegen) scTypeFromGo(typ types.Type) string {
+	switch t := typ.(type) {
 	case *types.Basic:
 		info := t.Info()
 		switch {

--- a/pkg/compiler/slice_test.go
+++ b/pkg/compiler/slice_test.go
@@ -1,10 +1,14 @@
 package compiler_test
 
 import (
+	"fmt"
 	"math/big"
+	"strings"
 	"testing"
 
+	"github.com/nspcc-dev/neo-go/pkg/compiler"
 	"github.com/nspcc-dev/neo-go/pkg/vm"
+	"github.com/stretchr/testify/require"
 )
 
 var sliceTestCases = []testCase{
@@ -173,6 +177,31 @@ var sliceTestCases = []testCase{
 
 func TestSliceOperations(t *testing.T) {
 	runTestCases(t, sliceTestCases)
+}
+
+func TestSliceEmpty(t *testing.T) {
+	srcTmpl := `package foo
+	func Main() int {
+		var a []int
+		%s
+		if %s {
+			return 1
+		}
+		return 2
+	}`
+	t.Run("WithNil", func(t *testing.T) {
+		src := fmt.Sprintf(srcTmpl, "", "a == nil")
+		_, err := compiler.Compile(strings.NewReader(src))
+		require.Error(t, err)
+	})
+	t.Run("WithLen", func(t *testing.T) {
+		src := fmt.Sprintf(srcTmpl, "", "len(a) == 0")
+		eval(t, src, big.NewInt(1))
+	})
+	t.Run("NonEmpty", func(t *testing.T) {
+		src := fmt.Sprintf(srcTmpl, "a = []int{1}", "len(a) == 0")
+		eval(t, src, big.NewInt(2))
+	})
 }
 
 func TestJumps(t *testing.T) {

--- a/pkg/compiler/struct_test.go
+++ b/pkg/compiler/struct_test.go
@@ -1,6 +1,7 @@
 package compiler_test
 
 import (
+	"fmt"
 	"math/big"
 	"testing"
 
@@ -365,4 +366,26 @@ var structTestCases = []testCase{
 
 func TestStructs(t *testing.T) {
 	runTestCases(t, structTestCases)
+}
+
+func TestStructCompare(t *testing.T) {
+	srcTmpl := `package testcase
+	type T struct { f int }
+	func Main() int {
+		a := T{f: %d}
+		b := T{f: %d}
+		if a != b {
+			return 2
+		}
+		return 1
+	}`
+	t.Run("Equal", func(t *testing.T) {
+		src := fmt.Sprintf(srcTmpl, 4, 4)
+		eval(t, src, big.NewInt(1))
+	})
+	t.Run("NotEqual", func(t *testing.T) {
+		src := fmt.Sprintf(srcTmpl, 4, 5)
+		eval(t, src, big.NewInt(2))
+	})
+
 }

--- a/pkg/compiler/struct_test.go
+++ b/pkg/compiler/struct_test.go
@@ -338,6 +338,29 @@ var structTestCases = []testCase{
 		}`,
 		big.NewInt(2),
 	},
+	{
+		"uninitialized struct fields",
+		`package foo
+		type Foo struct {
+			i int
+			m map[string]int
+			b []byte
+			a []int
+			s struct { ii int }
+		}
+		func NewFoo() Foo { return Foo{} }
+		func Main() int {
+			foo := NewFoo()
+			if foo.i != 0 { return 1 }
+			if len(foo.m) != 0 { return 1 }
+			if len(foo.b) != 0 { return 1 }
+			if len(foo.a) != 0 { return 1 }
+			s := foo.s
+			if s.ii != 0 { return 1 }
+			return 2
+		}`,
+		big.NewInt(2),
+	},
 }
 
 func TestStructs(t *testing.T) {

